simln-lib/src/lib.rs

```diff
@@ -1085,47 +1109,116 @@ impl<C: Clock + 'static> Simulation<C> {
+                                if c == payload.current_count {
+                                    log::info!(
+                                        "Payment count has been met for {}: {c} payments. Stopping the activity."
+                                    , executor.source_info);
+                                    continue
+                                }
+                            }
+                            let sender = producer_channels.get(&executor.source_info.pubkey).ok_or(
+                                SimulationError::RandomActivityError(RandomActivityError::ValueError(
+                                    format!(
+                                        "Activity producer for: {} not found.",
+                                        executor.source_info.pubkey,
+                                    ),
+                                )),
+                            )?;
+
+                            // pe: produce events
+                            let pe_shutdown = shutdown.clone();
+                            let pe_sender = sender.clone();
+                            let pe_clock = clock.clone();
+                            let source = executor.source_info.clone();
+                            // Only proceed with a payment if the amount is non-zero, otherwise skip this round. If we can't get
+                            // a payment amount something has gone wrong (because we should have validated that we can always
+                            // generate amounts), so we exit.
+                            let payment_amount = executor.payment_generator.payment_amount(payload.capacity);
+                            let amount = match payment_amount {
+                                Ok(amt) => {
+                                    if amt == 0 {
+                                        log::debug!(
+                                            "Skipping zero amount payment for {source} -> {}.", payload.destination
+                                        );
+                                        generate_payment(&mut heap, executor, payload.current_count, payment_event_payloads.clone()).await?;
+                                        continue;
+                                    }
+                                    amt
+                                },
+                                Err(e) => {
+                                    return Err(SimulationError::PaymentGenerationError(e));
+                                },
+                            };
+                            generate_payment(&mut heap, executor, payload.current_count + 1, payment_event_payloads.clone()).await?;
+
+                            // Wait until our time to next payment has elapsed then execute a random amount payment to a random
+                            // destination.
+                            pe_clock.sleep(payload.wait_time).await;
```

Comment:
Still need to select on shutdown?

Reply:
yes, otherwise the task that is processing the payment's head does not stop.

Reply:
yes you are right, it is necessary to add a `select` in this part